doc/parameter_server.md

```diff
@@ -0,0 +1,98 @@
+# Parameter Server Design
+
+## Overview
+
+
+A typical parameter server architecture contains three roles:
+
+- master, responsible for creating/deleting/scheduling pservers and workers
+- pserver, responsible for provide parameter pull/push/optimize/checkpoint service
+- worker, responsible for training minibatches of data
+
+
+Please refer to:
+
+![parameter_server](./images/parameter_server.png)
+
+## Key concepts
+
+
+### Parameter sharding
+
+We usually shard big parameters of a neural network model to some nodes under a rule, in order to achieve network load banlance. There are several kinds of parameter to be handled seperately:
+
+- Very big embedding table: Embedding table is a collection of <item id, embedding vector> pairs. There is also a mapping from item id to pserver id, so the corresonding embedding vector will be stored at the certain pserver
+
+- Big dense tensor: If the dense tensor parameter exceeds certain size, it will be sliced into several subtensors. The parameter name combines subtensor id will become a unique key, and the value is the subtensor(Please note that the slice tensor operation is zero-copy)
+
+- Small dense tensor: The small dense tensor parameter will be stored at certain pserver wholely
+
+### Parameter Initialization
+
+
+Parameter initialization of very big embedding table is lazy. For example, in online learing, there could be unkown item id in the training data. So, after worker send the unkown item id to pserver, pserver will initialize corresponding embedding vector and send back to worker. This is a get_or_create semantic.
+
+Other parameters could be initialized before training.
+
+
+## Workflow
+
+### Assumption
+
+- only support asynchronous SGD
+- only support worker failover, do not support pserver
```

Comment:
do not support pserver -> do not support pserver failover

Reply:
Done.